src/python/pants/core/util_rules/distdir.py

```diff
@@ -21,6 +20,11 @@ class DistDir:
     relpath: Path
 
 
+def is_child_of(path: Path, directory: Path) -> bool:
+    abs_path = path if path.is_absolute() else directory.joinpath(path).resolve()
```

Comment:
Hm this technically breaks the "no side effects" thing with the engine by calling `.resolve()`. It's good this is now isolated to one place at least so that we're less likely to keep using it.

Reply:
I spent a bit of time looking to see if there was a Python stdlib function that did this check for us already, but didn't find a convenient one.

Reply:
My concern is more using `Path().resolve()`. Ideally this function would solely use `PurePath` and not `Path` so that it didn't do any file operations. Those FS operations break the contract with the Rules API to avoid side effects.

This was a problem already before, so it's not a blocker. I'm not sure what the fix is in general, and this PR is good that it isolates this issue to only one call site.

Reply:
This should probably be implemented in terms of `fast_relpath_optional`.